dvc/stage/run.py

```diff
@@ -145,48 +170,109 @@ def run_stage(
     run(stage, dry=dry, checkpoint_func=checkpoint_func, run_env=run_env)
 
 
-@contextmanager
-def checkpoint_monitor(stage, callback_func, proc, killed):
-    if not callback_func:
-        yield None
-        return
-
-    logger.debug(
-        "Monitoring checkpoint stage '%s' with cmd process '%d'",
-        stage,
-        proc.pid,
-    )
-    done = threading.Event()
-    monitor_thread = threading.Thread(
-        target=_checkpoint_run,
-        args=(stage, callback_func, done, proc, killed),
-    )
-
-    try:
-        monitor_thread.start()
-        yield monitor_thread
-    finally:
-        done.set()
-        monitor_thread.join()
-
-
-def _checkpoint_run(stage, callback_func, done, proc, killed):
-    """Run callback_func whenever checkpoint signal file is present."""
-    signal_path = os.path.join(stage.repo.tmp_dir, CHECKPOINT_SIGNAL_FILE)
+@dataclass
```

Comment:
Looks like we need a scheduler already for the runner. :)

Reply:
I was just saying that all the logic in the `stage/run` does not belong here. 

We should think of decoupling all of the logic here and in the `repro` to a `Runner` and `TaskScheduler` respectively. `Runner` should support hooks (`on_start`/`on_end`/`on_failure`/`every(1s)` etc.) and actually running the stages.
And, the `TaskScheduler` that schedules stages to be run from the given DAG (which would help us parallelize in the future).

We'd likely come to it later though, just sharing my thoughts 🙂 . But this also looks good for now.

Reply:
@skshetry That sounds like a great idea!